rust/serde_bser/src/de/bunser.rs

```diff
@@ -0,0 +1,137 @@
+//! Internal stateless code for handling BSER deserialization.
+
+use byteorder::{ByteOrder, NativeEndian};
+
+use de::read::{DeRead, Reference};
+use errors::*;
+use header::*;
+
+#[derive(Debug)]
+pub struct Bunser<R> {
+    read: R,
+    scratch: Vec<u8>,
+}
+
+pub struct PduInfo {
+    pub bser_capabilities: u32,
+    len: i64,
+    start: i64,
+}
+
+impl<'de, R> Bunser<R>
+    where R: DeRead<'de>
+{
+    pub fn new(read: R) -> Self {
+        Bunser {
+            read,
+            scratch: Vec::with_capacity(128),
+        }
+    }
+
+    /// Read the PDU off the stream. This should be called in the beginning.
+    pub fn read_pdu(&mut self) -> Result<PduInfo> {
+        {
+            let magic = self.read_bytes(2)?;
+            if magic.get_ref() != &EMPTY_HEADER[..2] {
+                bail!("invalid magic header {:?}", magic);
+            }
+        }
+        let bser_capabilities = self.read.next_u32(&mut self.scratch)?;
+        let len = self.check_next_int()?;
+        let start = self.read_count();
+        Ok(PduInfo {
+               bser_capabilities,
+               len,
+               start,
+           })
+    }
+
+    pub fn read_count(&self) -> i64 {
+        self.read.read_count() as i64
+    }
+
+    pub fn end(&self, pdu_info: &PduInfo) -> Result<()> {
+        self.read
+            .verify_read_count((pdu_info.start + pdu_info.len) as usize)
+    }
+
+    #[inline]
+    pub fn peek(&mut self) -> Result<u8> {
+        self.read.peek()
+    }
+
+    #[inline]
+    pub fn discard(&mut self) {
+        self.read.discard();
+    }
+
+    /// Return a borrowed or copied version of the next n bytes.
+    #[inline]
+    pub fn read_bytes<'s>(&'s mut self, len: i64) -> Result<Reference<'de, 's, [u8]>> {
+        // TODO: what if the length can't be converted to a usize?
```

Comment:
It's probably reasonable to panic; it would imply either len is negative, or usize is 32 bits and len is > 2^32.